File: reader/Models/10 Demo/Population.R
###########################################
# Translation of Vensim file.
# Date created: 2018-08-08 14:45:20
###########################################
library(deSolve)
library(ggplot2)
library(tidyr)
#Displaying the simulation run parameters
START_TIME <- 1960.000000
FINISH_TIME <- 2010.000000
TIME_STEP <- 0.125000
#Setting aux param to NULL
auxs<-NULL

#Generating the simulation time vector
simtime<-seq(1960.000000,2010.000000,by=0.125000)
# Writing global variables (stocks and dependent auxs)
InitialPopulation = 3000
stocks <-c( Population = InitialPopulation )
# This is the model function called from ode
model <- function(time, stocks, auxs){
  with(as.list(c(stocks, auxs)),{
    GrowthFraction <- 0.015
    NumberAdded <- GrowthFraction*Population
    d_DT_Population  <- NumberAdded
    return (list(c(d_DT_Population)))
  })
}
# Function call to run simulation
o<-data.frame(ode(y=stocks,times=simtime,func=model,parms=auxs,method='euler'))
to<-gather(o,key=Stock,value=Value,2:ncol(o))
ggplot(data=to)+geom_line(aes(x=time,y=Value,colour=Stock))
#----------------------------------------------------
# Original text file exported from Vensim
#  Initial Population = 3000
#  Population = INTEG( Number Added , Initial Population ) 
#  FINAL TIME = 2010
#  Growth Fraction = 0.015
#  INITIAL TIME = 1960
#  Number Added = Growth Fraction * Population 
#  TIME STEP = 0.125
#----------------------------------------------------

File: reader/Models/10 Demo/Population.py
"""
Python model "Population.py"
Translated using PySD version 0.9.0
"""
from __future__ import division
import numpy as np
from pysd import utils
import xarray as xr

from pysd.py_backend.functions import cache
from pysd.py_backend import functions

_subscript_dict = {}

_namespace = {
    'TIME': 'time',
    'Time': 'time',
    'Growth Fraction': 'growth_fraction',
    'Initial Population': 'initial_population',
    'Number Added': 'number_added',
    'Population': 'population',
    'FINAL TIME': 'final_time',
    'INITIAL TIME': 'initial_time',
    'SAVEPER': 'saveper',
    'TIME STEP': 'time_step'
}

__pysd_version__ = "0.9.0"


@cache('run')
def growth_fraction():
    """
    Real Name: b'Growth Fraction'
    Original Eqn: b'0.015'
    Units: b''
    Limits: (None, None)
    Type: constant

    b''
    """
    return 0.015


@cache('run')
def initial_population():
    """
    Real Name: b'Initial Population'
    Original Eqn: b'3000'
    Units: b''
    Limits: (None, None)
    Type: constant

    b''
    """
    return 3000


@cache('step')
def number_added():
    """
    Real Name: b'Number Added'
    Original Eqn: b'Growth Fraction*Population'
    Units: b''
    Limits: (None, None)
    Type: component

    b''
    """
    return growth_fraction() * population()


@cache('step')
def population():
    """
    Real Name: b'Population'
    Original Eqn: b'INTEG ( Number Added, Initial Population)'
    Units: b''
    Limits: (None, None)
    Type: component

    b''
    """
    return integ_population()


@cache('run')
def final_time():
    """
    Real Name: b'FINAL TIME'
    Original Eqn: b'2010'
    Units: b'Year'
    Limits: (None, None)
    Type: constant

    b'The final time for the simulation.'
    """
    return 2010


@cache('run')
def initial_time():
    """
    Real Name: b'INITIAL TIME'
    Original Eqn: b'1960'
    Units: b'Year'
    Limits: (None, None)
    Type: constant

    b'The initial time for the simulation.'
    """
    return 1960


@cache('step')
def saveper():
    """
    Real Name: b'SAVEPER'
    Original Eqn: b'TIME STEP'
    Units: b'Year'
    Limits: (0.0, None)
    Type: component

    b'The frequency with which output is stored.'
    """
    return time_step()


@cache('run')
def time_step():
    """
    Real Name: b'TIME STEP'
    Original Eqn: b'0.125'
    Units: b'Year'
    Limits: (0.0, None)
    Type: constant

    b'The time step for the simulation.'
    """
    return 0.125


integ_population = functions.Integ(lambda: number_added(), lambda: initial_population())
